controllers/dataprotection/utils.go

```diff
@@ -149,58 +149,103 @@ func GetTargetPods(reqCtx intctrlutil.RequestCtx,
 	if !existPodSelector(selector) {
 		return nil, nil
 	}
-	labelSelector, err := metav1.LabelSelectorAsSelector(selector.LabelSelector)
-	if err != nil {
-		return nil, err
-	}
-	pods := &corev1.PodList{}
-	if err = cli.List(reqCtx.Ctx, pods,
-		client.InNamespace(reqCtx.Req.Namespace),
-		client.MatchingLabelsSelector{Selector: labelSelector}); err != nil {
-		return nil, err
+
+	listPodsByLabel := func(labelSelector *metav1.LabelSelector) (*corev1.PodList, error) {
+		if labelSelector == nil {
+			return nil, intctrlutil.NewNotFound("not found labelSelector")
+		}
+		selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+		if err != nil {
+			return nil, err
+		}
+		pods := &corev1.PodList{}
+		err = cli.List(reqCtx.Ctx, pods,
+			client.InNamespace(reqCtx.Req.Namespace),
+			client.MatchingLabelsSelector{Selector: selector})
+		return pods, err
 	}
 
-	if len(pods.Items) == 0 {
-		return nil, fmt.Errorf("failed to find target pods by backup policy %s/%s",
-			backupPolicy.Namespace, backupPolicy.Name)
+	pods, err := listPodsByLabel(selector.LabelSelector)
+	if err != nil {
+		return nil, err
 	}
-	sort.Sort(intctrlutil.ByPodName(pods.Items))
+	findPodErr := fmt.Errorf("failed to find target pods by backup policy %s/%s",
+		backupPolicy.Namespace, backupPolicy.Name)
 	var targetPods []*corev1.Pod
 	if len(selectedPodNames) == 0 || backupType == dpv1alpha1.BackupTypeContinuous {
 		switch selector.Strategy {
 		case dpv1alpha1.PodSelectionStrategyAny:
 			// always selecting the first pod
 			pod := dputils.GetFirstIndexRunningPod(pods)
+			if pod == nil {
+				pods, err := listPodsByLabel(selector.AlternateLabelSelector)
+				if err != nil && !intctrlutil.IsNotFound(err) {
+					return nil, err
+				}
+				if err == nil {
+					pod = dputils.GetFirstIndexRunningPod(pods)
+				}
+			}
 			if pod != nil {
 				targetPods = append(targetPods, pod)
 			}
 		case dpv1alpha1.PodSelectionStrategyAll:
+			allAvailable := true
+			for i := range pods.Items {
+				if !intctrlutil.IsAvailable(&pods.Items[i], 0) {
+					allAvailable = false
+					break
+				}
+			}
+			if len(pods.Items) == 0 || !allAvailable {
+				podList, err := listPodsByLabel(selector.AlternateLabelSelector)
+				if err != nil && !intctrlutil.IsNotFound(err) {
+					return nil, err
+				}
+				if err == nil {
+					pods = podList
+				}
+			}
+			if len(pods.Items) == 0 {
+				return nil, findPodErr
+			}
+			sort.Sort(intctrlutil.ByPodName(pods.Items))
 			for i := range pods.Items {
 				targetPods = append(targetPods, &pods.Items[i])
 			}
 		}
 		return targetPods, nil
 	}
+
 	// if already selected target pods and backupType is not Continuous, we should re-use them.
+	podMap := make(map[string]*corev1.Pod)
+	for i := range pods.Items {
+		podMap[pods.Items[i].Name] = &pods.Items[i]
+	}
+	pods, err = listPodsByLabel(selector.AlternateLabelSelector)
```

Comment:
why concatenate  pods selected by `alternateLabelSelector` with the ones selected by `labelSelector` into podMap?

Reply:
This is a better way to filter target pods. I'll change it.